File: sdk/ts/src/output/formatters/table.ts
/**
 * @module output/formatters/table
 *
 * Built-in table formatter. Hand-rolled aligner (matches the original
 * output.ts behaviour). Column order comes from the ColumnSpec list, or the
 * first row's key order when none was supplied; `cols` narrows and reorders.
 */

import type { Formatter, Options } from '../formatter';
import { deriveHeaders } from '../projection';

export const tableFormatter: Formatter = {
  key: 'table',
  extensions: [],
  options: [],
  render(out, data, _opts: Options, cols) {
    const rows = normalise(data);
    // Emptiness is a ROW-count decision: no rows means no output at all,
    // not a bare header line.
    if (rows.length === 0) return;

    // `cols` arrives pre-resolved from dispatch; empty means payload keys.
    const headers = cols.length > 0 ? cols : deriveHeaders(rows);
    if (headers.length === 0) return;

    const cells = rows.map(row =>
      headers.map(h => String(row[h] ?? '')),
    );
    const widths = headers.map((h, ci) =>
      Math.max(h.length, ...cells.map(r => r[ci].length)),
    );
    const pad = (s: string, w: number) => s + ' '.repeat(w - s.length);
    const line = (parts: readonly string[]) =>
      parts.map((c, i) => pad(c, widths[i])).join('  ');

    out.write(line(headers) + '\n');
    for (const row of cells) {
      out.write(line(row) + '\n');
    }
  },
};

function normalise(v: unknown): Record<string, unknown>[] {
  if (Array.isArray(v)) return v as Record<string, unknown>[];
  return [v as Record<string, unknown>];
}

File: sdk/experimental/rs/src/output/builtins/table.rs
//! Plain ASCII table formatter — the default --format.
//!
//! Mirrors the cli-php TableFormatter: header line + space-separated
//! body, no borders, no color, pipe-/grep-friendly. Adopters wanting
//! richer tables (borders, UTF-8 box-drawing, theming) override the
//! 'table' key in the Registry with their own Formatter.
//!
//! Renderer: built on `comfy-table::Preset::NOTHING` so column widths
//! are computed for us and we don't reinvent padding math.
//!
//! Options:
//!   - `header` (bool, default true) — emit the header row.

use std::io::Write;

use comfy_table::presets::NOTHING;
use comfy_table::{Cell, ContentArrangement, Row, Table};
use serde_json::Value;

use crate::output::formatter::Formatter;
use crate::output::option::{OptionSpec, OptionType, OptionValue, Options};

use super::rows::{normalize, resolve_columns, row_get, stringify};

static OPTS: &[OptionSpec] = &[OptionSpec {
    name: "header",
    r#type: OptionType::Bool,
    usage: "Emit a header row (default: true)",
    default: None, // merged at render-time
    r#enum: &[],
}];

pub struct TableFormatter;

impl Formatter for TableFormatter {
    fn key(&self) -> &'static str {
        "table"
    }

    fn extensions(&self) -> &'static [&'static str] {
        &[]
    }

    fn options(&self) -> &'static [OptionSpec] {
        OPTS
    }

    fn render(
        &self,
        out: &mut dyn Write,
        data: &Value,
        opts: &Options,
        cols: &[String],
    ) -> std::io::Result<()> {
        let header = opts
            .get("header")
            .and_then(OptionValue::as_bool)
            .unwrap_or(true);
        let rows = normalize(data);

        // Zero rows emits nothing — not even a bare header row. Emptiness is
        // decided by ROW count, never by header count: `cols` is non-empty
        // whenever the caller supplied a ColumnSpec list, so guarding on the
        // column count would print a lone header for an empty payload.
        if rows.is_empty() {
            return Ok(());
        }

        let columns = resolve_columns(&rows, cols);

        let mut table = Table::new();
        table
            .load_preset(NOTHING)
            .set_content_arrangement(ContentArrangement::Disabled);

        if header && !columns.is_empty() {
            table.set_header(columns.iter().map(Cell::new).collect::<Vec<_>>());
        }

        for row in &rows {
            let cells: Vec<Cell> = columns
                .iter()
                .map(|c| Cell::new(stringify(row_get(row, c))))
                .collect();
            table.add_row(Row::from(cells));
        }

        writeln!(out, "{}", table)?;
        Ok(())
    }
}
